include/drivers/gpio.h

```diff
@@ -30,14 +30,59 @@ extern "C" {
  * @brief GPIO Driver APIs
  * @defgroup gpio_interface GPIO Driver APIs
  * @ingroup io_interfaces
+ *
+ * @{
+ * @name GPIO Atomicity and Interrupt-Related Behavior
+ *
  * @{
+ * All GPIO peripheral API operations satisfy the following promises:
+ *
+ * * All functions shall be thread-safe and interrupt-safe.
+ * * All functions may be invoked from interrupt service routines and
+ *   callbacks.  Functions that cannot fulfill their behavioral
+ *   requirements in that context may return `-EWOULDBLOCK`.
+ * * All pin, interrupt, and callback configuration functions must be
```

Comment:
Current implementation of callback configuration functions (which is common to all drivers) is not thread safe, neither atomic. Someone needs to implement the feature. Should be likely in the same PR.